spi/ssd1309_spi/ssd1309_spi.c

```diff
@@ -0,0 +1,144 @@
+/**
+ * Copyright (c) 2025 mjcross
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+**/
+
+// A simple example to display text on an ssd1309-based OLED display with
+// an SPI interface. It should also work on a ssd1306 device (not tested).
+//
+// To understand the commands and addressing modes for the display refer to
+// the manufacturer's datasheet at https://www.hpinfotech.ro/SSD1309.pdf
+
+
+#include <stdio.h>
+#include "pico/stdlib.h"
+#include "hardware/spi.h"
+#include "ctype.h"
+#include "ssd1309_font.h"
+
+
+// We will use the spi0 peripheral and the following GPIO pins (see the
+// GPIO function select table in the RPi Pico datasheet).
+#define SPI_DEVICE      spi0
+#define PIN_CS          17      // chip select (active low)
+#define PIN_SCK         18      // SPI clock
+#define PIN_MOSI        19      // SPI data transmit (MOSI)
+#define PIN_DC          20      // data/command mode (low for command)
+#define PIN_R           21      // reset (active low)
+
+// ssd1309 accepts a maximum SPI clock rate of 10 Mbit/sec 
+#define SPI_BITRATE     10 * 1000 * 1000
+
+// dimensions of the display
+#define NUM_X_PIXELS    128
+#define NUM_Y_PIXELS    64
+
+// modes for the data/command pin
+#define DC_COMMAND_MODE 0
+#define DC_DATA_MODE    1
+
+
+// send a command byte to the display
+void send_command(uint8_t cmd_byte) {
+    gpio_put(PIN_DC, DC_COMMAND_MODE);
+    spi_write_blocking(SPI_DEVICE, &cmd_byte, 1);
+}
+
+// send a data byte to the display
+void send_data(uint8_t data_byte) {
+    gpio_put(PIN_DC, DC_DATA_MODE);
+    spi_write_blocking(SPI_DEVICE, &data_byte, 1);
+}
+
+// set the text cursor position (row, col)
+void set_cursor_pos(uint text_row, uint text_col) {
+    send_command(0xb0 + (text_row & 0x07));     // set the text row ('page') start address
+    uint col = text_col * 8;
+    send_command(col & 0x0f);                   // set the column start address (low nibble)
+    send_command(0x10 + ((col & 0xf0) >> 4));   // set the column start address (high nibble)
+}
+
+// display a single text character at the cursor position
+void display_char(char c) {
+    uint font_index = 0;             // default (space)
+    c = toupper(c);
+    if (isalpha(c)) {
```

Comment:
Minor nitpick: I guess you could put the `c = toupper(c);` inside `if (isalpha(c)) {` :wink:

Reply:
Fair enough.